Firestore/core/src/firebase/firestore/immutable/sorted_map_iterator.h

```diff
@@ -0,0 +1,186 @@
+      : tag_(Tag::Tree), tree_iter_(std::move(delegate)) {
+  }
+
+  SortedMapIterator(const SortedMapIterator& other) : tag_(other.tag_) {
+    switch (tag_) {
+      case Tag::Array:
+        new (&array_iter_) ArrayIter(other.array_iter_);
+        break;
+      case Tag::Tree:
+        new (&tree_iter_) TreeIter(other.tree_iter_);
+        break;
+    }
+  }
+
+  SortedMapIterator(SortedMapIterator&& other) : tag_(other.tag_) {
+    switch (tag_) {
+      case Tag::Array:
+        new (&array_iter_) ArrayIter(std::move(other.array_iter_));
+        break;
+      case Tag::Tree:
+        new (&tree_iter_) TreeIter(std::move(other.tree_iter_));
+        break;
+    }
+  }
+
+  ~SortedMapIterator() {
+    switch (tag_) {
+      case Tag::Array:
+        array_iter_.~ArrayIter();
+        break;
+      case Tag::Tree:
+        tree_iter_.~TreeIter();
+        break;
+    }
+  }
+
+  SortedMapIterator& operator=(const SortedMapIterator& other) {
+    if (tag_ == other.tag_) {
+      switch (tag_) {
+        case Tag::Array:
+          array_iter_ = other.array_iter_;
+          break;
+        case Tag::Tree:
+          tree_iter_ = other.tree_iter_;
+          break;
+      }
+    } else {
+      this->~SortedMapIterator();
+      new (this) SortedMapIterator(other);
+    }
+    return *this;
+  }
+
+  SortedMapIterator& operator=(SortedMapIterator&& other) {
+    if (tag_ == other.tag_) {
+      switch (tag_) {
+        case Tag::Array:
+          array_iter_ = std::move(other.array_iter_);
+          break;
+        case Tag::Tree:
+          tree_iter_ = std::move(other.tree_iter_);
+          break;
+      }
+    } else {
+      this->~SortedMapIterator();
+      new (this) SortedMapIterator(std::move(other));
+    }
+    return *this;
+  }
+
+  pointer get() const {
+    switch (tag_) {
+      case Tag::Array:
+        return array_iter_;
+      case Tag::Tree:
+        return tree_iter_.get();
+    }
+  }
+
+  reference operator*() const {
+    return *get();
+  }
+
+  pointer operator->() const {
+    return get();
+  }
+
+  SortedMapIterator& operator++() {
+    switch (tag_) {
+      case Tag::Array:
+        ++array_iter_;
+        break;
+      case Tag::Tree:
+        ++tree_iter_;
+        break;
+    }
+    return *this;
+  }
+
+  SortedMapIterator operator++(int /*unused*/) {
+    SortedMapIterator result = *this;
+    ++*this;
+    return result;
+  }
+
+  friend bool operator==(const SortedMapIterator& a,
+                         const SortedMapIterator& b) {
+    if (a.tag_ != b.tag_) {
+      return false;
+    }
+
+    switch (a.tag_) {
+      case Tag::Array:
+        return a.array_iter_ == b.array_iter_;
+      case Tag::Tree:
+        return a.tree_iter_ == b.tree_iter_;
+    }
+  }
+
+  bool operator!=(SortedMapIterator b) const {
+    return !(*this == b);
+  }
+
+ private:
+  enum class Tag {
+    Array,
+    Tree,
+  };
+
+  Tag tag_;
+  union {
+    ArrayIter array_iter_;
```

Comment:
I guess as an alternative to having this class be parametrized by iterator types, it could be parametrized by just `K, V`:
```
typename impl::FixedArray<std::pair<K, V>>::const_iterator array_iter_;
typename impl::LlrbNode<K, V>::const_iterator tree_iter_;
```
Entirely optional. This would make the class somewhat more constrained (as it stands, it's a tagged union of two arbitrary iterators), which can be seen as a good or as a bad thing.